### PR TITLE
Rework monitoring actor to use local state instead of tasks

### PR DESCRIPTION
**Pull request by thomaseizinger**
This should greatly improve resilience as we no longer have any
tasks within `tokio` that can fail without recovering.

Instead, we periodically ping the actor with a `Sync` message which
updates the local state of all scripts and sends out messages accordingly.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
Will rebase on top of https://github.com/comit-network/hermes/pull/145 once merged.

File: daemon/src/monitor.rs [hunk 1]
```diff
@@ -23,41 +29,46 @@ pub struct MonitorParams {
...
-                .map(|(tx, _, range)| (tx.txid(), script_pubkey.clone(), range))
-                .collect(),
-            refund: (
-                dlc.refund.0.txid(),
-                script_pubkey,
-                refund_timelock_in_blocks,
-            ),
-        }
-    }
+pub struct Sync;
```
Comment (da-kami) [inline daemon/src/monitor.rs:32]: ❓ So this message is used to trigger syncing from the outside and configured when instantiating the actor?

(just validating I got it correctly 😬)

**Review by da-kami — COMMENTED** [review]
Looking good, but needs are rebase on latest master - can't test

The log messages changed, I don't see the transaction monitoring being picked up for existing Cfds Open anymore, but I suppose that is expected.

File: daemon/src/monitor.rs (hunk 1 above)
Comment (da-kami) [reply daemon/src/monitor.rs:32]: read the f***ing PR description Daniel!

File: daemon/src/monitor.rs [hunk 2]
```diff
@@ -279,30 +224,200 @@ where
...
+        // 4. check for finished monitoring tasks
+        for ((txid, script), status) in self.current_status.iter() {
+            match self.awaiting_status.entry((*txid, script.clone())) {
+                Entry::Vacant(_) => {
+                    unreachable!("we are only fetching the status of scripts we are waiting for")
+                }
+                Entry::Occupied(mut occupied) => {
+                    let targets = occupied.insert(Vec::new());
+
+                    // Split vec into two lists, all the once for which we reached the target and
```
Comment (da-kami) [inline daemon/src/monitor.rs]:
```suggestion
                    // Split vec into two lists, all the ones for which we reached the target and
```

File: daemon/src/monitor.rs [hunk 3]
```diff
@@ -279,30 +224,200 @@ where
...
+                    // the once which we need to still monitor
```
Comment (da-kami) [inline daemon/src/monitor.rs]:
```suggestion
                    // the ones which we need to still monitor
```

File: daemon/src/monitor.rs [hunk 4]
```diff
@@ -279,30 +224,200 @@ where
...
+
+impl Confirmed {
+    fn with_confirmations(blocks: u32) -> Self {
+        Self { depth: blocks - 1 }
+    }
+
+    /// Compute the depth of a transaction based on its inclusion height and the
+    /// latest known block.
+    ///
+    /// Our information about the latest block might be outdated. To avoid an
```
Comment (da-kami) [inline daemon/src/monitor.rs:359]: thanks for the comment, I started thinking about the saturating sub again...

File: daemon/src/monitor.rs [hunk 5]
```diff
@@ -525,3 +529,275 @@ where
...
+                            height: *included_at,
+                            tx_hash: *tx,
+                            fee: None,
+                        }]
+                    })
+                    .collect())
+            }
+
+            fn batch_call(&self, _batch: &Batch) -> Result<Vec<serde_json::Value>, Error> {
+                todo!()
```
Comment (da-kami) [inline daemon/src/monitor.rs]: nit: maybe use `unreachable` instead? I always feel `todo` indicates something that should be actioned at some point

**Review by da-kami — APPROVED** [review]
I gave it a more thorough review and did not spot any issues. However, I think only using this for some time will reveal bugs. 
The test is cool, interesting concepts in there.

File: daemon/src/monitor.rs (hunk 5 above)
Comment (thomaseizinger) [reply daemon/src/monitor.rs]: This was just autofilled by rust-analyzer but yeah, I can change it to unreachable!

File: daemon/src/monitor.rs (hunk 4 above)
Comment (thomaseizinger) [reply daemon/src/monitor.rs:359]: This comment is a copy paste from the old code lol :D

**Pull request merged**